File: src/main/java/gg/mineral/practice/util/collection/ProfileList.java
package gg.mineral.practice.util.collection;

import gg.mineral.api.collection.GlueList;
import gg.mineral.practice.entity.Profile;
import lombok.Getter;
import lombok.val;

import java.io.Serial;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;

@Getter
public class ProfileList extends ConcurrentLinkedQueue<Profile> {

    /**
     *
     */
    @Serial
    private static final long serialVersionUID = 1L;

    public ProfileList() {
        super();
    }

    public ProfileList(Collection<? extends Profile> c) {
        super(c);
    }

    @Override
    public boolean add(Profile profile) {
        if (contains(profile))
            return false;
        return super.add(profile);
    }

    public boolean contains(Profile p) {

        if (p == null)
            return false;

        for (Profile profile : this)
            if (p.getUuid().equals(profile.getUuid()))
                return true;

        return false;
    }

    public List<Profile> subList(int fromIndex, int toIndex) {
        if (fromIndex < 0 || toIndex > size() || fromIndex > toIndex)
            throw new IndexOutOfBoundsException();

        List<Profile> subList = new GlueList<>();
        Iterator<Profile> iterator = iterator();

        for (int i = 0; i < fromIndex; i++) {
            if (iterator.hasNext()) {
                iterator.next();
            } else {
                throw new IndexOutOfBoundsException();
            }
        }

        for (int i = fromIndex; i < toIndex; i++) {
            if (iterator.hasNext()) {
                subList.add(iterator.next());
            } else {
                throw new IndexOutOfBoundsException();
            }
        }

        return subList;
    }

    public Profile removeFirst() {
        return poll();
    }

    public Profile getFirst() {
        return peek();
    }

    public Profile get(UUID uuid) {
        // TODO: Optimize this using hashmaps
        for (val profile : this)
            if (profile.getUuid().equals(uuid))
                return profile;

        return null;
    }
}

File: src/main/kotlin/gg/mineral/practice/util/collection/ProfileList.kt
package gg.mineral.practice.util.collection

import gg.mineral.api.collection.GlueList
import gg.mineral.practice.entity.Profile
import java.io.Serial
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

class ProfileList : ConcurrentLinkedQueue<Profile> {
    constructor() : super()

    constructor(c: Collection<Profile>) : super(c)

    override fun add(element: Profile): Boolean {
        if (contains(element)) return false
        return super.add(element)
    }

    override fun contains(element: Profile): Boolean {
        for (profile in this) if (element.uuid == profile.uuid) return true
        return false
    }

    fun subList(fromIndex: Int, toIndex: Int): List<Profile> {
        if (fromIndex < 0 || toIndex > size || fromIndex > toIndex) throw IndexOutOfBoundsException()

        val subList: MutableList<Profile> = GlueList()
        val iterator: Iterator<Profile> = iterator()

        fromIndex.rangeTo(0)
            .forEach { _ -> if (iterator.hasNext()) iterator.next() else throw IndexOutOfBoundsException() }

        fromIndex.rangeTo(toIndex)
            .forEach { _ -> if (iterator.hasNext()) subList.add(iterator.next()) else throw IndexOutOfBoundsException() }

        return subList
    }

    fun removeFirst(): Profile? = poll()

    val first: Profile?
        get() = peek()

    fun get(uuid: UUID): Profile? {
        // TODO: Optimize this using hashmaps
        for (profile in this) if (profile.uuid == uuid) return profile

        return null
    }

    companion object {
        /**
         *
         */
        @Serial
        private const val serialVersionUID = 1L
    }
}
